File: Windows/ChipBurner/ChipBurner/Hex/HexParser.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Hex
{
	public class HexParser
	{
		private const int DATALENOFFSET = 1;
		private const int DATAOFFSET = 9;
		private const int ADDRESSOFFSET = 3;
		private const int TYPEOFFSET = 7;

		private string FilePath { get; set; }
		private StreamReader File { get; set; }
		private int Endianess { get; set; }
		public int LastAddressOffset = 0;

		public HexParser(string filePath,int endianness = 0)
		{
			FilePath = filePath;
			Endianess = endianness;
		}

		public void StartReading()
		{
			LastAddressOffset = 0;
			if (File == null)
				File = new StreamReader(FilePath);
			else
				File.BaseStream.Seek(0, SeekOrigin.Begin);
		}

		public HexRecord ReadLine()
		{
			HexRecord record = new HexRecord();
			if (File.EndOfStream)
			{
				record.Type = HexRecord.EOF;
				return record;
			}

			string line = File.ReadLine();

			record.Type = RecordType(line);
			string data = null;
			switch(record.Type){
				case HexRecord.DataType:
					data = RecordData(line);

					record.ExtendedAddress = LastAddressOffset;
					record.Address = HexToWord(RecordDataAddress(line));
					record.Length = RecordDataLength(line);
					record.CheckSum = 0;//not used
					record.Data = new byte[record.Length];

					for (int i = 0; i < data.Length; i += 2)//two because it's in Hex characters, 2 character = 1 byte
						record.Data[i / 2] = (byte)((HexToByte(data[i]) << 4) | HexToByte(data[i + 1]));
					break;
				case HexRecord.EOF:
					break;
				case HexRecord.ExtendedSegmentAddress:
					break;
				case HexRecord.ExtendedLinearAddressRecord:
					data = RecordData(line);
					record.Length = RecordDataLength(line);
					record.Data = new byte[record.Length];
					for (int i = 0; i < data.Length; i += 2)//two because it's in Hex characters, 2 character = 1 byte
						record.Data[i / 2] = (byte)((HexToByte(data[i]) << 4) | HexToByte(data[i + 1]));
					record.ExtendedAddress = (record.Data[0] << 8) | record.Data[1];
					LastAddressOffset = record.ExtendedAddress << 16;
					break;
					
			}
			return record;
		}

		private string RecordDataAddress(string block)
		{
			return block.Substring(ADDRESSOFFSET, 4);
		}

		private byte HexToByte(char hex)
		{
			return HexConverter.HexToByte(hex);
		}

		private int HexToWord(string hex)
		{
			return (int)HexConverter.HexToMax4Bytes(hex);
		}

		private int RecordType(string block)
		{
			int type = HexToByte(block[TYPEOFFSET]) * 16 + HexToByte(block[TYPEOFFSET + 1]);
			switch (type)
			{

				case HexRecord.DataType:
					return HexRecord.DataType;
				case HexRecord.EOF:
					return HexRecord.EOF;
				case HexRecord.ExtendedSegmentAddress:
					return HexRecord.ExtendedSegmentAddress;
				case HexRecord.ExtendedLinearAddressRecord:
					return HexRecord.ExtendedLinearAddressRecord;
				default:
					throw new Exception("Record Type Unknown!");
			}
		}

		private int RecordDataLength(string block)
		{
			return HexToByte(block[DATALENOFFSET]) * 16 + HexToByte(block[DATALENOFFSET + 1]);
		}

		private string RecordData(string block)
		{
			int type = RecordType(block);

			if (type == HexRecord.EOF)
				return null;

			int len = RecordDataLength(block);
			string ret = block.Substring(DATAOFFSET, 2 * len);

			return ret;
		}

		public virtual void Dispose()
		{
			File.Close();
			File.Dispose();
		}
	}
}

File: IDE/Windows/LeoninoUSBJava/src/hex/HexParser.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package hex;

import java.io.*;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 *
 * @author rudarobson
 */
public class HexParser {

    private static int DATALENOFFSET = 1;
    private static int DATAOFFSET = 9;
    private static int ADDRESSOFFSET = 3;
    private static int TYPEOFFSET = 7;
    private String FilePath;
    private BufferedReader File;
    private int Endianess;
    public int LastAddressOffset = 0;

    public HexParser(String filePath) {
        this(filePath, 0);
    }

    public HexParser(String filePath, int endianness) {
        FilePath = filePath;
        Endianess = endianness;
    }

    public void StartReading() throws FileNotFoundException {
        LastAddressOffset = 0;
        if (File == null) {
            File = new BufferedReader(new FileReader(FilePath));
        }
    }

    public HexRecord ReadLine() {
        HexRecord record = new HexRecord();
//        if (File.EndOfStream) {
//            record.Type = HexRecord.EOF;
//            return record;
//        }
        try {
            String line = File.readLine();

            record.setType(RecordType(line));
            String data = null;
            switch (record.getType()) {
                case HexRecord.DataType:
                    data = RecordData(line);

                    record.setAddress(LastAddressOffset);
                    record.setAddress(HexToWord(RecordDataAddress(line)));
                    record.setLength(RecordDataLength(line));
                    record.setCheckSum(0);
                    record.setData(new byte[record.getLength()]);

                    for (int i = 0; i < data.length(); i += 2)//two because it's in Hex characters, 2 character = 1 byte
                    {
                        record.getData()[i / 2] = (byte) ((HexToByte(data.charAt(i)) << 4) | HexToByte(data.charAt(i + 1)));
                    }

                    break;
                case HexRecord.EOF:
                    break;
                case HexRecord.ExtendedSegmentAddress:
                    break;
                case HexRecord.ExtendedLinearAddressRecord:
                    data = RecordData(line);
                    record.setLength(RecordDataLength(line));
                    record.setData(new byte[record.getLength()]);
                    for (int i = 0; i < data.length(); i += 2)//two because it's in Hex characters, 2 character = 1 byte
                    {
                        record.getData()[i / 2] = (byte) ((HexToByte(data.charAt(i)) << 4) | HexToByte(data.charAt(i + 1)));
                    }


                    record.setExtendedAddress((record.getData()[0] << 8) | record.getData()[1]);
                    LastAddressOffset = record.getExtendedAddress() << 16;
                    break;

            }
        } catch (Exception ex) {
        }
        return record;
    }

    private String RecordDataAddress(String block) {
        return block.substring(ADDRESSOFFSET, ADDRESSOFFSET + 4);
    }

    private byte HexToByte(char hex) {
        return HexConverter.HexToByte(hex);
    }

    private int HexToWord(String hex) {
        return (int) HexConverter.HexToMax4Bytes(hex);
    }

    private int RecordType(String block) {
        int type = HexToByte(block.charAt(TYPEOFFSET)) * 16 + HexToByte(block.charAt(TYPEOFFSET + 1));
        switch (type) {

            case HexRecord.DataType:
                return HexRecord.DataType;
            case HexRecord.EOF:
                return HexRecord.EOF;
            case HexRecord.ExtendedSegmentAddress:
                return HexRecord.ExtendedSegmentAddress;
            case HexRecord.ExtendedLinearAddressRecord:
                return HexRecord.ExtendedLinearAddressRecord;
            default:
                throw new RuntimeException("Record Type Unknown!");
        }
    }

    private int RecordDataLength(String block) {
        return HexToByte(block.charAt(DATALENOFFSET)) * 16 + HexToByte(block.charAt(DATALENOFFSET + 1));
    }

    private String RecordData(String block) {
        int type = RecordType(block);

        if (type == HexRecord.EOF) {
            return null;
        }

        int len = RecordDataLength(block);
        String ret = block.substring(DATAOFFSET, DATAOFFSET + 2 * len);

        return ret;
    }

    public void Dispose() {
        try {
            File.close();
        } catch (IOException ex) {
            throw new RuntimeException();
        }
    }
}
